src/noise.rs

```diff
@@ -31,3 +33,22 @@ pub fn rms(s: Signal, period: TimeSecs) -> f64 {
         })
         .sqrt()
 }
+
+/// Applies Gaussian white noise to a Signal at the specified signal to noise ratio.
+///
+/// The function uses the Additive Gaussian white noise model
+/// https://en.wikipedia.org/wiki/Additive_white_Gaussian_noise
+///
+/// # Arguments
+///
+/// * `s` - The base Signal to apply noise to.
+/// * `signal_to_noise` - The desired signal to noise ratio for the returned signal.
+/// * `sample_duration` - A period to analyse the base Signal (should be around one period of the signal)
+pub fn gaussian_white_noise(s: Signal, signal_to_noise: f64, sample_duration: TimeSecs) -> Signal {
+    let rms_signal = rms(s.clone(), sample_duration);
```

Comment:
It seems like I should change the `rms` function and even the `gaussian_white_noise` functions signal arguments to something else to avoid having to `clone`. I only need to read the signal.

Reply:
Imo it's OK to not space-optimize signals yet. We don't create very many of them and have at most tens of them around a time.

Rust gives us opportunities to explore these side quests and it's hard to resist. I was able to shave some clones by having sample take a reference to a signal rather than owning it. It took me a while to figure out how to promise the borrow checker that the iterator wouldn't outlive the signal:

```rust
pub fn sample<'s>(rate: FrequencyHz, s: &'s Signal) -> impl Iterator<Item = f64> + 's {
    (0..).map(move |n: u32| {
        let sample_period = rate.period();
        let t = n as f64 * sample_period.0;
        s.at(TimeSecs(t))
    })
}

pub fn rms(s: &Signal, period: TimeSecs) -> f64 {
     // unchanged
}
```

The cost of the `.clone()` is the cost of cloning the `Arc` (just an atomic reference count bump) and copying over the numeric fields in the `Signal` struct. My "optimization" of keeping extra information around in the struct may not have been optimal 🤷‍♂️.

Reply:
Thanks ! When I tried I couldn't figure out the borrow annotation part, now it's clear.